Delete the sample in v4 and update the readme (#7)
Azure-Samples/app-service-msi-keyvault-python: README.md

# How to set and get secrets from Azure Key Vault with Azure Managed Identities and Python

## This sample shows how to do the following operations of Key Vault secret with Key Vault SDK
- Get Key Vault MSIAuthentication or ServicePrincipalCredentials
- Create a Key Vault client
- Get an existing secret

## Background

> For service to service authentication, the approach involved creating an Azure AD application and associated credential, and using that credential to get a token. While this approach works well, there are two shortcomings:

Reviewer: Regarding line 41: clicking the link (with the corrected `href`) will take you to a page with a couple of required fields that should maybe be mentioned in the sample. Users have to choose/create a resource group, select a region, specify a website name, and specify a Key Vault name. The current wording suggests that only a secret value needs to be specified.

> For service to service authentication, the approach involved creating an Azure AD application and associated credential, and using that credential to get a token. While this approach works well, there are two shortcomings:

Reviewer: Regarding line 34: as I mentioned in the main conversation tab, the `href` value for the link should be changed to `"https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure-Samples%2Fazure-sdk-for-python-keyvault-secrets-get-set-managedid%2Fmaster%2Fv4%2Fazuredeploy.json"`

> For service to service authentication, the approach involved creating an Azure AD application and associated credential, and using that credential to get a token. While this approach works well, there are two shortcomings:

Reviewer 2: Regarding line 41: there are other required fields when deploying the template, and we might want to mention them as well (resource group, website name, key vault name). This might suggest to a user that only the secret needs to be specified.

> For service to service authentication, the approach involved creating an Azure AD application and associated credential, and using that credential to get a token. While this approach works well, there are two shortcomings:

Reviewer: Regarding line 56: the "Add New" button has been replaced with an "Add Access Policy" button, so this should be updated.

2. [Azure CLI 2.0] to run the application on your local development machine.

### Step 1: Create an App Service with an Azure Managed Identity
<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure-Samples%2Fazure-sdk-for-python-keyvault-secrets-get-set-managedid%2Fmaster%2Fazuredeploy.json" target="_blank">
    <img src="http://azuredeploy.net/deploybutton.png"/>
</a>

Use the "Deploy to Azure" button to deploy an ARM template to create the following resources:
1. App Service with [Azure Managed Identities].
2. Key Vault with a secret, and an access policy that grants the App Service access to **Get Secrets**.

> >Note: Clicking the link will take you to a page with a couple of required fields that should maybe be mentioned in the sample. Users have to choose/create a resource group, select a region, specify a website name, and specify a Key Vault name. The current wording suggests that only a secret value needs to be specified.

Reviewer: Sorry, I wasn't clear on what should go here. How about:
```suggestion
>Note: When preparing the deployment, there will be a few required fields to fill out (subscription, resource group, region, website name, Key Vault name, and secret value). The secret value will be the value of the secret named "secret" created in the Key Vault upon deployment.
```

>     git clone https://github.com/Azure-Samples/azure-sdk-for-python-keyvault-secrets-get-set-managedid.git

Reviewer: Just to be clear that the following commands need to be run from inside the directory
```suggestion
    git clone https://github.com/Azure-Samples/azure-sdk-for-python-keyvault-secrets-get-set-managedid.git
    cd azure-sdk-for-python-keyvault-secrets-get-set-managedid
```

>     git clone https://github.com/Azure-Samples/azure-sdk-for-python-keyvault-secrets-get-set-managedid.git

Reviewer 2: To make it clear that the following commands need to be run from inside the directory:
```suggestion
    git clone https://github.com/Azure-Samples/azure-sdk-for-python-keyvault-secrets-get-set-managedid.git
    cd azure-sdk-for-python-keyvault-secrets-get-set-managedid
```

    ```

4.  Run the following command to install dependencies:

    ```
    pip install -r requirements.txt
    ```

> 5.  Set up the environment variable `KEY_VAULT_URL` with your KeyVault URL or replace the variable in the example file.

Reviewer: `examply.py` attempts to get `KEY_VAULT_URI` from the environment, so this will need to align with that:
```suggestion
5.  Set up the environment variable `KEY_VAULT_URI` with your KeyVault URI or replace the variable in the example file.
```

> 

Reviewer: Maybe an example would be useful here:
```
SET KEY_VAULT_URI=https://{your vault name}.vault.azure.net/  # setting environment variable in Windows command prompt
```

> 

Reviewer 2: It may be worth adding an example here:
```
SET KEY_VAULT_URI="https://{your vault name}.vault.azure.net/"  # set environment variable in Windows command prompt
```

   - `/ping` : This just answers "hello world" and is a good way to test if your packages are installed correctly without testing Azure itself.

>    - `/` : The MSI sample itself

Reviewer: After setting the environment variables, making sure my service principal has correct permissions, and confirming that a secret named "secret" exists in my vault, this endpoint only brings up an error with the message "vaultBaseUrl". I've tried debugging this for a bit but I'm not sure where the issue is.

Reviewer: Resolved -- I had set the vault URL environment variable inside quotations. Removing them made the sample run as expected.

## Use latest Key Vault SDK

The Key Vault SDK package version here is **0.3.x**, if you are using the [latest](https://pypi.org/project/azure-keyvault-secrets/) version of the Key Vault secret SDK package, please refer to the following examples:

 * [helloworld.py](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/keyvault/azure-keyvault-secrets/samples/hello_world.py) - Examples for common Key Vault secret tasks:

   * Get DefaultAzureCredential
   * Create a secret client
   * Create a new secret
   * Get an existing secret
   * Update an existing secret
   * Delete a secret

## Deploying on Azure Web App

> 1. Set the `KEY_VAULT_URL` environment variable using the "Application Settings" of your Web App.

Reviewer: This may need to be `KEY_VAULT_URI` as well.
